util/sharding_writer/sharding_writer.go

```diff
@@ -0,0 +1,316 @@
+	if w.opts.shardSize <= 0 {
+		// 2^64 bytes in a file ought to be enough for anybody.
+		w.opts.shardSize = math.MaxInt64
+	}
+
+	var err error
+	w.currentFile, err = os.Create(w.currentFileName())
+	if err != nil {
+		log.Fatalf("unable to create sharded file %q", w.currentFileName())
+	}
+	return w
+}
+
+// currentFileName is a helper to format the current filename.
+func (w *Writer) currentFileName() string {
+	// TODO(rsned): Come to a decision on shard number width and padding
+	// that will match what the Python library does.
+	// E.g., %d, %05d, etc.
+	return fmt.Sprintf("%s_%d%s", w.basePath, w.currentShard, w.extension)
+}
+
+// Write writes len(p) bytes from p to the underlying data stream. It returns
+// the number of bytes written from p (0 <= n <= len(p)) and any error encountered
+// that caused the write to stop early.
+func (w *Writer) Write(data []byte) (n int, err error) {
+	if w.errorTriggered {
+		return 0, fmt.Errorf("error occurred during previous writes")
+	}
+
+	var bytesWritten int64
+	for len(data) > 0 {
+		// Check if we need to start a new shard.
+		if w.shardFull() {
+			if err := w.newShard(); err != nil {
+				w.errorTriggered = true
+				return int(bytesWritten), err
+			}
+		}
+		// Figure out how much of the input we can use before sharding.
+		splitPoint := w.opts.shardSize - w.bytesWritten
+
+		// If we don't have enough to fill out the remainder of
+		// the current shard, adjust the cut point.
+		if splitPoint > int64(len(data)) {
+			splitPoint = int64(len(data))
+		}
+
+		n, err = w.currentFile.Write(data[:splitPoint])
+		bytesWritten += int64(n)
+		w.bytesWritten += int64(n)
+		if (err != nil) || int64(n) != splitPoint {
+			w.errorTriggered = true
+			return int(bytesWritten), fmt.Errorf("error writing: %v", err)
+		}
+
+		// Drop the portion written so far.
+		data = data[splitPoint:]
+	}
+
+	return int(bytesWritten), nil
+}
+
+// WriteString writes the contents of the given string. It returns the number n of
+// bytes written from from the string (0 <= n <= len(s)) and any error encountered
+// that caused the write to stop early.
+func (w *Writer) WriteString(s string) (n int, err error) {
+	if w.errorTriggered {
+		return 0, fmt.Errorf("error occurred during previous writes")
+	}
+
+	// track total bytes written in this call, which may be more than shardSize.
+	var bytesWritten int64
+
+	switch w.opts.dataType {
+	case dataTypeBytes:
+		for len(s) > 0 {
+			// Check if we need to start a new shard.
+			if w.shardFull() {
+				if err := w.newShard(); err != nil {
+					w.errorTriggered = true
+					return int(bytesWritten), err
+				}
+			}
+			// Figure out how much of the input we can use before sharding.
+			splitPoint := w.opts.shardSize - w.bytesWritten
+
+			// If we don't have enough to fill out the remainder of
+			// the current shard, adjust the cut point.
+			if splitPoint > int64(len(s)) {
+				splitPoint = int64(len(s))
+			}
+
+			n, err = w.currentFile.WriteString(s[:splitPoint])
```

Comment:
How does this function differ from Writer.Write except for this function call?  Is it possible to not repeat the code?  Or does Golang's type system and lack of polymorphism prevent that?